File: travamento.py
import threading
import time

THREAD_BAIXA = "Thread Baixa"
THREAD_MEDIA = "Thread Media"
THREAD_ALTA = "Thread Alta"

barramento_ocupado = False
mutex = threading.Lock()
not_busy = threading.Condition(mutex)

cpu_ocupada = False

def registrar_log(nome_thread, msg):
    print(f"[{time.monotonic()}] [{nome_thread}]: {msg}\n")

def exibir_tempo_total(nome_thread, inicio):
    tempo_gasto = time.monotonic() - inicio
    print(f"\n>>> [RESULTADO] A {nome_thread} levou {tempo_gasto:.4f} segundos do inicio ao fim. <<<\n")


def adquirir_barramento(nome_thread):
    global barramento_ocupado
    not_busy.acquire()

    while barramento_ocupado:
        registrar_log(nome_thread, "Barramento ocupado. Thread vai esperar...")
        not_busy.wait()

    barramento_ocupado = True
    registrar_log(nome_thread, "Adquiriu o barramento com sucesso.")

    not_busy.release()


def liberar_barramento(nome_thread):
    global barramento_ocupado
    not_busy.acquire()

    barramento_ocupado = False
    registrar_log(nome_thread, "Liberou o barramento.")
    not_busy.notify()

    not_busy.release()


def thread_baixa():
    global cpu_ocupada
    inicio = time.monotonic()
    registrar_log(THREAD_BAIXA, "Iniciando...")
    adquirir_barramento(THREAD_BAIXA)

    registrar_log(THREAD_BAIXA, "Realizando escrita lenta no barramento (Dividida em etapas)...")

    for _ in range(5):
        while cpu_ocupada:
            registrar_log(THREAD_BAIXA, "CPU está ocupada. Aguardando liberação.")
            time.sleep(0.05)
        time.sleep(0.1)  # Trabalho real da etapa

    liberar_barramento(THREAD_BAIXA)
    registrar_log(THREAD_BAIXA, "Finalizada.")
    exibir_tempo_total(THREAD_BAIXA, inicio)


def thread_media():
    global cpu_ocupada
    time.sleep(0.05)  # Acorda logo para pegar a CPU
    inicio = time.monotonic()

    registrar_log(THREAD_MEDIA, "Iniciando. PREEMPTANDO a Baixa e Monopolizando a CPU...")

    cpu_ocupada = True

    contador = 0
    # Aumentado para 50 milhões para gerar um gargalo de tempo perceptível
    for i in range(50000000):
        contador += 1

    cpu_ocupada = False
    registrar_log(THREAD_MEDIA, "Finalizou calculos intensivos. Devolveu a CPU.")
    exibir_tempo_total(THREAD_MEDIA, inicio)


def thread_alta():
    time.sleep(0.02)
    inicio = time.monotonic()

    registrar_log(THREAD_ALTA, "Iniciando requisicao critica de tempo real...")
    adquirir_barramento(THREAD_ALTA)

    registrar_log(THREAD_ALTA, "Enviando dados...")
    time.sleep(0.1)
    liberar_barramento(THREAD_ALTA)

    registrar_log(THREAD_ALTA, "Finalizada.")
    exibir_tempo_total(THREAD_ALTA, inicio)


def main():
    print("--- INICIANDO SIMULADOR MARS PATHFINDER (O TRAVAMENTO) ---\n")

    tb = threading.Thread(target=thread_baixa)
    ta = threading.Thread(target=thread_alta)
    tm = threading.Thread(target=thread_media)

    tb.start()
    ta.start()
    tm.start()

    tb.join()
    ta.join()
    tm.join()


if __name__ == "__main__":
    main()

File: travamento.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <time.h>
#include <unistd.h>

typedef struct {
    int barramento_ocupado;
    pthread_mutex_t mutex;
    pthread_cond_t cond; 
} Monitor;

Monitor monitor_barramento;

// Variável que simula a existência de apenas 1 processador (como na sonda)
int cpu_monopolizada = 0; 

void registrar_log(const char* nome_thread, const char* acao) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts); 
    printf("[%ld.%09ld] [%s]: %s\n", ts.tv_sec, ts.tv_nsec, nome_thread, acao);
}

void exibir_tempo_total(const char* nome_thread, struct timespec inicio) {
    struct timespec fim;
    clock_gettime(CLOCK_MONOTONIC, &fim);
    double tempo_gasto = (fim.tv_sec - inicio.tv_sec) + (fim.tv_nsec - inicio.tv_nsec) / 1e9;
    printf("\n>>> [RESULTADO] A %s levou %.4f segundos do inicio ao fim. <<<\n\n", nome_thread, tempo_gasto);
}

void adquirir_barramento(const char* nome_thread) {
    pthread_mutex_lock(&monitor_barramento.mutex);
    while (monitor_barramento.barramento_ocupado == 1) {
        registrar_log(nome_thread, "Barramento ocupado. Entrando em estado bloqueado (Wait)...");
        pthread_cond_wait(&monitor_barramento.cond, &monitor_barramento.mutex);
    }
    monitor_barramento.barramento_ocupado = 1;
    registrar_log(nome_thread, "Adquiriu o barramento com sucesso.");
    pthread_mutex_unlock(&monitor_barramento.mutex);
}

void liberar_barramento(const char* nome_thread) {
    pthread_mutex_lock(&monitor_barramento.mutex);
    monitor_barramento.barramento_ocupado = 0;
    registrar_log(nome_thread, "Liberou o barramento.");
    pthread_cond_signal(&monitor_barramento.cond);
    pthread_mutex_unlock(&monitor_barramento.mutex);
}

void* thread_baixa(void* arg) {
    struct timespec inicio; clock_gettime(CLOCK_MONOTONIC, &inicio);
    registrar_log("Thread Baixa", "Iniciando...");
    adquirir_barramento("Thread Baixa");
    
    registrar_log("Thread Baixa", "Realizando escrita lenta no barramento (Dividida em etapas)...");
    
    // Simula a escrita lenta em 5 etapas
    for(int i = 0; i < 5; i++) {
        // Se a Thread Média tomou a CPU, a Baixa fica congelada aqui (Simulação de Preempção)
        while(cpu_monopolizada == 1) {
            usleep(50000); 
        }
        usleep(100000); // Trabalho real da etapa
    }
    
    liberar_barramento("Thread Baixa");
    registrar_log("Thread Baixa", "Finalizada.");
    exibir_tempo_total("Thread Baixa", inicio);
    return NULL;
}

void* thread_media(void* arg) {
    usleep(100000); // Espera a Baixa e a Alta começarem
    struct timespec inicio; clock_gettime(CLOCK_MONOTONIC, &inicio);

    registrar_log("Thread Media", "Iniciando. PREEMPTANDO a Baixa e Monopolizando a CPU...");
    cpu_monopolizada = 1; // Tira a CPU da Thread Baixa
    
    volatile long contador = 0;
    for(long i = 0; i < 2000000000L; i++) { contador++; }
    
    cpu_monopolizada = 0; // Devolve a CPU
    registrar_log("Thread Media", "Finalizou calculos intensivos. Devolveu a CPU.");
    exibir_tempo_total("Thread Media", inicio);
    return NULL;
}

void* thread_alta(void* arg) {
    usleep(50000); 
    struct timespec inicio; clock_gettime(CLOCK_MONOTONIC, &inicio);

    registrar_log("Thread Alta ", "Iniciando requisicao critica de tempo real...");
    adquirir_barramento("Thread Alta ");
    
    registrar_log("Thread Alta ", "Enviando dados...");
    usleep(100000); 
    liberar_barramento("Thread Alta ");
    
    registrar_log("Thread Alta ", "Finalizada.");
    exibir_tempo_total("Thread Alta ", inicio);
    return NULL;
}

int main() {
    pthread_t tb, tm, ta;
    monitor_barramento.barramento_ocupado = 0;
    pthread_mutex_init(&monitor_barramento.mutex, NULL);
    pthread_cond_init(&monitor_barramento.cond, NULL);
    
    printf("--- INICIANDO SIMULADOR MARS PATHFINDER (O TRAVAMENTO) ---\n\n");
    pthread_create(&tb, NULL, thread_baixa, NULL);
    pthread_create(&ta, NULL, thread_alta, NULL);
    pthread_create(&tm, NULL, thread_media, NULL);
    
    pthread_join(tb, NULL); pthread_join(ta, NULL); pthread_join(tm, NULL);
    return 0;
}